Grad of int

LenaMartens opened the pull request:
It is currently not possible to take a gradient with respect to integer inputs in Jax. When an integer in- or output is encountered a hard error is thrown. The aim of this PR is to support integers in AD by returning a zero-vector in a trivial vector space when taking the gradient with respect to an integer. More concretely: if the primal value is integer or boolean, the tangent value will be a zero-vector with a trivial dtype (`float0`). 

Steps:

- [x] Add new `float0` dtype representing a trivial vector space, along with a `Float0Array` to represent the arrays that will be returned as tangent values 
- [x] Support `jvp` of the identity function with integer primal and `float0` tangent
- [x] Ignore `float0` args in jvp rules (either by replacing them with `ad.Zero` at the jvp boundary or special casing the rules) and test jvp of non-identity functions
- [x]  Support `grad` of integers
- [x] Support `jit` in presence of `float0`s
- [x] Make all tests pass 
- [x] Add method to cast `Float0Array`s to float zero-vectors 
- [x] Add a clear error explaining `float0`s and differentiating wrt int/bool inputs. Ideally it would be an error on any unsupported operation on the `Float0Array` (and maybe where the float0 originated?).

LenaMartens commented on jax/core.py:
```diff
@@ -751,7 +751,7 @@ class AbstractValue:
 
   def at_least_vspace(self):
-    assert False
+    return self
```
I'm calling `at_least_vspace` in a few places now (like `JVPTrace`) to instantiate tangent avals of the correct dtype, but this results in sometimes hitting the `assert False` in `AbstractValue` here. As a quick fix this now does `return self`. 
I'm not sure what the `assert False` is guarding against: maybe it doesn't make sense to reason about the vspace of an `AbstractValue` because it shouldn't have a tangent value? Any objection to removing the assert?

mattjj replied:
I think it was meant to be something like a "raise NotImplementedError", or like an abstract method on an abstract base class. Using `return self` could be a sensible default, though I'm interested in what errors pop up if we leave the `assert False`, just to investigate whether there are other solutions...

LenaMartens replied:
For example in `api_test` this happens in ``

```
```

LenaMartens replied:
For example in `api_test` the assertion is hit in CustomJVPTest.test_initial_style, CustomVJPTest.test_initial_style and RematTest.test_remat_symbolic_zeros.  In these tests it hits a `full_raise` on a `JVPTrace`. I think before this change the `full_raise` will return a `JVPTrace` with a `Zero(AbstractValue)` as a tangent.

I've left in the assert if you want to test it out.

End of the CustomJVPTest.test_initial_style stacktrace:

```
 File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/interpreters/partial_eval.py", line 139, in process_primitive
    return custom_partial_eval_rules[primitive](self, *tracers, **params)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/lax/lax_control_flow.py", line 1593, in _scan_partial_eval
    unroll=unroll)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/lax/lax_control_flow.py", line 1829, in scan_bind
    return core.Primitive.bind(scan_p, *args, **params)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/core.py", line 276, in bind
    tracers = map(top_trace.full_raise, args)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/util.py", line 35, in safe_map
    return list(map(f, *args))
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/core.py", line 372, in full_raise
    return self.pure(val)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/interpreters/ad.py", line 243, in pure
    tangent_zero = Zero(get_aval(val).at_least_vspace())
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/core.py", line 754, in at_least_vspace
    assert False
AssertionError
```

LenaMartens commented on jax/lax/lax.py:
```diff
@@ -2451,6 +2451,8 @@ def _convert_element_type_translation_rule(c, operand, *, new_dtype, old_dtype):
 
 def _convert_element_type_transpose_rule(t, *, new_dtype, old_dtype):
   assert t.dtype == new_dtype, (t.dtype, new_dtype)
+  if core.primal_dtype_to_tangent_dtype(old_dtype) == dtypes.float0:
+    return [ad_util.Zero(core.Float0Array(t.shape))]
```
This special case was necessary to fix the `test_vjp_of_int_fulllike ` test. Without this fix that test fails with an assert failure here: https://github.com/google/jax/blob/04f9a7e53db3eb0eb6d5d08a5f2a967bf9712048/jax/interpreters/ad.py#L158. This happened in quite a few places because some `lax` primitives use `full_like` with an integer constant.

My working theory is this only happens when tracking the tangent of a ConstVar in the `backward_pass`, and this tangent not matching the cotangent generated by traversing the Jaxpr backwards through a `convert_element_type`.  It only happens when the tangent is a known `PartialValue`, otherwise the `float0` cotangent is wrapped in a `Zero` and correctly ignored. I think (?) the `PartialValue` is only known when the variable we are tracking is a ConstVar, but I'm very unsure there.

It feels like maybe there is a better fix here: should we be tracking the tangent of ConstVars in the first place? Is there a more central place to guard against this "TypeError" than in the transpose rules?

mattjj replied:
IIRC we think this might be an issue with `convert_element_type`'s JVP rule not being correct (i.e. we shouldn't treat it as linear, since e.g. the JVP rule for `lambda x: convert_element_type(x, old_dtype=float32, new_dtype=int32)` should be `lambda x, t: (convert_element_type(x, old_dtype=float32, new_dtype=int32), convert_element_type(t, old_dtype=float32, new_dtype=float0)` (or if we prefer the second element of the return value should be a symbolic zero with float0 as its aval's dtype).

Is that right?

LenaMartens replied:
yes! have now updated the jvp rule of  `convert_element_type` to return symbolic zeros when `new_dtype` is int/bool and the transpose_rule to do the same for `old_dtype`, I _think_ that's how we discussed it.

mattjj commented on jax/api.py:
```diff
@@ -1780,7 +1784,11 @@ def fun(*tangents):
 
 def _check_inexact_input_vjp(x):
```
Should we just get rid of this check entirely? On one hand it seems like we're expanding the scope of autodiff to handle exact dtypes; on the other hand it might catch some accidental errors if people accidentally write `grad(f)(0)` when they mean `grad(f)(0.)`.

My current hunch is we should remove this check.

LenaMartens replied:
I think it's a common enough mistake to use integer arguments where you intended floats, and it now errors clearly while after this change it will return these `float0`s which require some (non-trivial) context.
Ideally there would be an error on any unsupported operation on the `Float0Array` explaining you differentiated wrt an integer argument. I like the idea of adding the source line of where the array was created (I think I've seen such errors already in context of omnistaging, so would be cool to replicate that)

Is there a way to override all errors when trying to use an unimplemented method or should I add it for the most commonly used ones?

mattjj commented on jax/core.py:
```diff
@@ -909,7 +909,9 @@ def __repr__(self):
 
   def at_least_vspace(self) -> AbstractValue:
-    return self
+    # TODO(lenamartens): maybe defensive, where do UnshapedArrays come in?
```
This looks good to me!

I think UnshapedArrays never arise nowadays. We haven't removed them because I think they're conceptually nice, but we never made use of them because tracing at an unshaped level is not very useful.

mattjj commented on jax/core.py:
```diff
@@ -1040,6 +1044,13 @@ def strip_weak_type(self) -> 'ConcreteArray':
+  # TODO(lenamartens): maybe make a mapping here instead of if, although there
+  # might only be two cases now where we need to return a different dtype.
+  # Question. How to do robust dtype comparison?
+  if primal_dtype == np.int32 or primal_dtype == np.int64 or primal_dtype == np.bool:
```
How about `if not dtypes.issubdtype(primal_dtype, np.inexact): ...`?

mattjj commented:
This is looking really good! I checked through everything except the xla.py changes, which might need to be tweaked a little for compatibility with #3983 if that one goes in first.

cgarciae commented:
Quick question: what would `int + float0` cast to?

mattjj commented:
@cgarciae  I don't think we plan to support automatic dtype promotion, or even necessarily to support any jax.numpy functions with this dtype. So you won't be able to do much with a float0 array! It's basically a dummy.

cgarciae commented:
Hmm. But if by mistake one of the parameters of the network is an `int32` and its gradient is of type `float0`, a library like optix/optax will try to add them up.

mattjj commented:
> But if by mistake one of the parameters of the network is an int32 and its gradient is of type float0

Yes, all the more reason not to silently promote the dtype, and instead require users be explicit about what they want! Users/libraries who differentiate with respect to integer dtypes should know what they're doing.

cgarciae commented:
It makes sense. It will be a bit weird for users however that the error appears in the optimizer code and not directly on `grad`, might get hard to debug.

Can `grad` be modified to give an error or warning via a flag?

mattjj commented:
Yeah, there's a comment I left on this PR today to that effect; on master, there is such an error.

But maybe we can do a good enough job of raising a good error when operating on a float0 array. We could even track that the value arose from a particular grad evaluation...

LenaMartens commented on jax/interpreters/ad.py:
```diff
@@ -296,13 +300,17 @@ def process_custom_jvp_call(self, _, __, f_jvp, tracers):
     primals_in, tangents_in = unzip2((t.primal, t.tangent) for t in tracers)
     primals_in = map(core.full_lower, primals_in)
     tangents_in = map(instantiate_zeros, tangents_in)
+    tangents_in = [aval.to_zeros() if type(aval) is core.Float0Array
```
The reason for adding this here is to fully instantiate the Float0Arrays as regular float zeros because they will be passed into the custom_jvp rules. They are not ignored, like they are on the regular jvp path. The reason for this is not entirely clear to me, and although this work-around makes my tests pass it does not feel like a true fix. Where on the custom_jvp path do we ignore the Zero tangents?

This particular edge case happens when calling a function that internally calls a custom_jvped function with an integer constant, like in nn.softplus here: https://github.com/google/jax/blob/3f8aaabbccb40bb5c1c69e1f15ba7bd30dd3fdbc/jax/nn/functions.py#L41.

mattjj replied:
I think the `aval` name might be misleading here, since the values in `tangents_in` aren't avals but just values. How about

```suggestion
    tangents_in = [t.to_zeros() if type(t) is core.Float0Array else t for t in tangents_in]
```

mattjj commented on jax/ad_util.py:
```diff
@@ -43,13 +44,19 @@ def add_impl(xs, ys):
+  return core.Float0Array(a.shape)
+
 jaxval_zeros_likers: Dict[type, Array] = {}
+jaxval_zeros_likers[dtypes.dtype(core.Float0Array)] = float0s_like
```
I think `jaxval_zeros_likers` should map from a Python type (i.e. class) to the zeros-like function, rather than from dtype to the zeros like function. For example, we expect `numpy.ndarray` to be a key in this table but not all the numpy dtypes. That is, I think we want something like

```suggestion
jaxval_zeros_likers[core.Float0Array] = float0s_like
```

mattjj commented on jax/ad_util.py:
```diff
@@ -43,13 +44,19 @@ def add_impl(xs, ys):
+
 jaxval_zeros_likers: Dict[type, Array] = {}
+jaxval_zeros_likers[dtypes.dtype(core.Float0Array)] = float0s_like
+jaxval_zeros_likers[dtypes.dtype(int)] = float0s_like
```
I don't think we need this one

```suggestion
```

mattjj commented on jax/core.py:
```diff
@@ -32,6 +32,7 @@
 from .config import FLAGS, config
 from . import linear_util as lu
 from . import source_info_util
+from .numpy import _operations, _nondiff_methods, diff_methods
```
We don't want a dependence on jax.numpy in core (I think it'll lead to circular import issues, and to the extent that we want some kind of dependence tree rather than every module importing every other module, we want core to be minimal).

Maybe this was stale code anyway?

```suggestion
```

mattjj commented on jax/ad_util.py:
```diff
@@ -43,13 +44,19 @@ def add_impl(xs, ys):
 
 aval_zeros_likers: Dict[type, Array] = {}
 aval_zeros_likers[AbstractUnit] = lambda _: unit
+aval_zeros_likers[core.Float0Array] = lambda aval: aval
```
I think the keys in this table should be AbstractValue subclasses (and the type annotation should be more precise). Since float0 arrays reuse existing abstract values rather than their own, we probably don't need to extend the table.

```suggestion
```

mattjj commented on jax/ad_util.py:
```diff
@@ -43,13 +44,19 @@ def add_impl(xs, ys):
 def zeros_like_aval(aval):
   return aval_zeros_likers[type(aval)](aval)
 
 aval_zeros_likers: Dict[type, Array] = {}
```
See comment below for why I think this type annotation wasn't precise enough to be helpful.

```suggestion
aval_zeros_likers: Dict[Type[core.AbstractValue], Array] = {}
```

LenaMartens replied:
Updated type annotation.

mattjj commented on jax/api.py:
```diff
@@ -1680,8 +1684,8 @@ def _jvp(fun: lu.WrappedFun, primals, tangents):
-    if _dtype(p) != _dtype(t):
-      msg = ("primal and tangent arguments to jax.jvp must have equal types; "
+    if core.primal_dtype_to_tangent_dtype(_dtype(p)) != _dtype(t):
+      msg = ("primal and tangent arguments to jax.jvp do not match; must be equal or float0 in case of int/bool. "
```
```suggestion
msg = ("primal and tangent arguments to jax.jvp do not match; "
       "dtypes must be equal, or in case of int/bool primal dtype "
       "the tangent dtype must be float0."
       f"Got primal dtype {_dtype(p)} and so expected tangent dtype "
       f"{core.primal_dtype_to_tangent_dtype(_dtype(p))}, but got "
       f"tangent dtype {_dtype(t)} instead.")
```

mattjj commented on jax/api.py:
```diff
@@ -1862,9 +1872,13 @@ def _vjp(fun: lu.WrappedFun, *primals, **kwargs):
+  ct_dtypes = [core.primal_dtype_to_tangent_dtype(_dtype(x)) for x in out_primal]
   vjp_py = Partial(partial(_vjp_pullback_wrapper,
-                           [_dtype(x) for x in out_primal],
+                           # cotangent dtypes
```
Do you want to keep these comments?

mattjj commented on jax/interpreters/ad.py:
```diff
@@ -296,13 +300,17 @@ def process_custom_jvp_call(self, _, __, f_jvp, tracers):
   def process_custom_vjp_call(self, _, __, fwd, bwd, tracers, *, out_trees):
     primals_in, tangents_in = unzip2((t.primal, t.tangent) for t in tracers)
     tangents_in = map(instantiate_zeros, tangents_in)
+    tangents_in = [aval.to_zeros() if type(aval) is core.Float0Array
```
Similar to the above:

```suggestion
    tangents_in = [t.to_zeros() if type(t) is core.Float0Array else t for t in tangents_in]
```

mattjj commented on jax/numpy/lax_numpy.py:
```diff
@@ -4583,6 +4584,27 @@ def _operator_round(number, ndigits=None):
+# Throw an error when trying to operate on Float0Arrays giving context about
+# what float0s are.
+def float0_error(*args, **kwargs):
+  # TODO(lenamartens): before merging: write a better error here :)
```
Don't forget this todo!

LenaMartens replied:
Done, PTAL :)

mattjj reviewed (commented):
Overall LGTM! A few comments below (plus some test failures to follow up on).

Great work!

LenaMartens commented:
I think all tests pass now!

jekbradbury commented:
Our new Copybara flow requires that any squashing we want be done manually (and it looks like this PR should probably be squashed); can you do that? (Or maybe we want a change to the CL export script?)

LenaMartens commented:
I've squashed the commits. The CL export script works correctly in that it always amends and force pushes one commit but I switched to a git workflow in the middle of the PR.

shoyer commented on jax/core.py:
```diff
@@ -1623,3 +1630,39 @@ def initial_style_staging():
     finally:
       trace_state.initial_style = prev
+
+class Float0Array:
+  """Array used to represent the tangent values of int/bool primals.
+
+  No operations are defined for this array-type by design: these values can't be
+  used in combination with non-trivial vector spaces. No implicit dtype
+  conversion is done.  To use these values, the array can be cast to float32
+  zeros using the `to_zeros` method.
+  """
```
I have mixed feelings about the `Float0Array` class. I understand the motivation to explicitly not support any operations, and the desire to raise errors as close as possible to the source of the issue as possible, but I am concerned that this special case may cause more trouble than it's worth.

One concern is type stability/predictable. Currently, the return value from functions transformed with `grad` and `jit` is always a pytree of `DeviceArray` objects (but see https://github.com/google/jax/issues/4386). Now it's `DeviceArray` and/or `Float0Array` objects.

In contrast, the standard way to add new dtypes in NumPy/JAX is to just add a new dtype and use the standard array object, relying on the lack of operation implementations (e.g., for arithmetic) to raise errors, but still supporting methods like `transpose`/`reshape`, which are otherwise valid on arrays with any dtype.

A good example of this is the `jax.jacrev`, which despite being well-defined currently results in an error from inside `_unravel_array_into_pytree`:
```
In [29]: jax.jacrev(lambda x: 1.0)(1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-29-d57c9bab827f> in <module>
----> 1 jax.jacrev(lambda x: 1.0)(1)

~/dev/jax/jax/api.py in jacfun(*args, **kwargs)
    975     jac = jac[0] if isinstance(argnums, int) else jac
    976     example_args = dyn_args[0] if isinstance(argnums, int) else dyn_args
--> 977     jac = tree_map(partial(_unravel_array_into_pytree, y, 0), jac)
    978     return tree_transpose(tree_structure(example_args), tree_structure(y), jac)
    979

~/dev/jax/jax/tree_util.py in tree_map(f, tree)
    162   """
    163   leaves, treedef = pytree.flatten(tree)
--> 164   return treedef.unflatten(map(f, leaves))
    165
    166 def tree_multimap(f, tree, *rest):

~/dev/jax/jax/api.py in _unravel_array_into_pytree(pytree, axis, arr)
   1060   axis = axis % arr.ndim
   1061   shapes = [arr.shape[:axis] + np.shape(l) + arr.shape[axis+1:] for l in leaves]
-> 1062   parts = _split(arr, np.cumsum(map(np.size, leaves[:-1])), axis)
   1063   reshaped_parts = [np.reshape(x, shape) for x, shape in zip(parts, shapes)]
   1064   return tree_unflatten(treedef, reshaped_parts)

~/dev/jax/jax/api.py in _split(x, indices, axis)
   1068     return np.split(x, indices, axis)
   1069   else:
-> 1070     return x.split(indices, axis)
   1071
   1072 def _dtype(x):

~/dev/jax/jax/numpy/lax_numpy.py in float0_error(self, *args, **kwargs)
   4657 def float0_error(self: core.Float0Array, *args, **kwargs):
   4658   raise TypeError(
-> 4659       "Float0Arrays do not support any operations by design, because float0s "
   4660       "are not compatible with non-trivial vector spaces. No implicit dtype "
   4661       "conversion is done. You can use to_zeros() to cast this array to "

TypeError: Float0Arrays do not support any operations by design, because float0s are not compatible with non-trivial vector spaces. No implicit dtype conversion is done. You can use to_zeros() to cast this array to float32 zeros.
If you didn't expect to get a Float0Array you might have accidentally taken a gradient with respect to an integer or boolean argument.
```

You can try to fix all these cases inside JAX, but it's a bit of an uphill battle. Sticking to operations that are unsupported by a dtype would be a bit simpler.

LenaMartens replied:
Thanks for the review Stephan, I agree this is not very neat currently. The only reason for having a separate array type was to provide more context on an error, but I think it's a good idea to do this at the dtype promotion level. It makes more sense in a way: what we're trying to prevent is implicit dtype casting because this could lead to silently doing the wrong thing. 

> the standard way to add new dtypes in NumPy/JAX is to just add a new dtype and use the standard array object

Which array object is this? DeviceArray? a numpy array? 

> As long as it's excluded from all dtype promotion, I imagine it could be perfectly fine (and maybe even useful) to support arbitrary operation like lax.add.

Would this mean only supporting this for float0+float0 cases because there is no promotion implemented?

LenaMartens replied:
I guess one additional difference is a Float0Array does not actually store any values, it only defines a shape. Is there a way to express this using the existing arrays? Maybe a zeros array with float0 values which are defined as taking up 0 bytes is actually the same as not storing values?

In this way it's like a ShapedArray, but @mattjj explained to me we need an actual array next to our abstract arrays, so I don't think we can reuse ShapedArray?

LenaMartens replied:
After some discussion we decided to abandon the Float0Array and opt for ndarrays with float0 dtype. This allows for dtype-agnostic operations (like reshape or split). The float0error is now thrown when promoting the dtype of the arguments in lax and lax.numpy.

shoyer commented on jax/core.py:
```diff
@@ -1623,3 +1630,39 @@ def initial_style_staging():
+  def to_zeros(self):
+    return np.zeros(self.shape, dtype=np.float32)
+
+  def __eq__(self, other) -> bool:
+    return type(self) is type(other) and self.shape == other.shape
```
Nit: this definition of equality is incompatible with the definition on other JAX/NumPy arrays.

shoyer commented on jax/core.py:
```diff
@@ -1623,3 +1630,39 @@ def initial_style_staging():
+
+  __array_priority__ = 100
+
+  # Float0Array has methods that are dynamically populated in lax_numpy.py,
+  # and this annotation is needed to make pytype happy.
+  _HAS_DYNAMIC_ATTRIBUTES = True
```
The downside of this approach is that now pytype can't catch these errors at compile time (but I guess it would have hard time identifying when `Float0Array` is created, anyways).

shoyer commented:
Other edge cases that came up in my testing:

1. Both `jacfwd` and `jacrev` raise errors.
2. `np.array(Float0Array)` returns a 0d object dtype array. If you decide to keep `Float0Array`, you should define a `__array__` method to fix this. It should probably either return a NumPy array with `float0` dtype or raise an error.

LenaMartens committed:
Support grad with integer arguments.
- Add float0 and set-up at_least_vspace to return float0
values for int/bool primals
- Use Zero to wrap float0 tangents so they're correctly ignored in jvp
rules
- Add float0 handlers to XLA to support jit
- Fix convert_element_type and tie_in jvp rules

LenaMartens commented:
Added a `allow_int` flag to grad so you have to opt-in to the grad-of-int behavior. For jvp and vjp no checks are done. This means most accidental use of integer arguments in a grad context will still be captured by the current input data-type checks.

mattjj commented on jax/api.py:
```diff
@@ -1940,15 +1952,17 @@ def make_jaxpr(fun: Callable,
       containers (tuple/list/dict) thereof.
+      arguments and return value should be arrays, scalars, or standard Python
     static_argnums: See the :py:func:`jax.jit` docstring.
+      containers (tuple/list/dict) thereof.
```
Is there a typo around here, maybe switched lines?

LenaMartens replied:
yep!

mattjj commented on jax/core.py:
```diff
@@ -1581,3 +1589,6 @@ def initial_style_staging():
       trace_state.initial_style = prev
+
+# Casting float0 array to a float-valued zero array.
+zeros_like_float0 = lambda a: np.zeros(a.shape, np.float)
```
```suggestion
def zeros_like_float0(a): return np.zeros(a.shape, np.float)
```

Optional change, but it can be slightly nicer in debuggers to have a function name.

mattjj commented on jax/lax/lax.py:
```diff
@@ -1944,11 +1944,21 @@ def naryop_dtype_rule(result_dtype, accepted_dtypes, name, *avals, **kwargs):
+            "conversion is done. You can use np.zeros_like(arr, dtype=np.float) "
+            "to cast a float0 array to a regular zeros array. \n"
+            "If you didn't expect to get a float0 you might have accidentally "
+            "taken a gradient with respect to an integer argument.")
```
Very nice!

mattjj commented on jax/lib/xla_bridge.py:
```diff
@@ -289,6 +289,8 @@ def supported_numpy_dtypes():
   """Normalize dtypes in a value."""
   if hasattr(val, '__array__') or np.isscalar(val):
+    if dtypes.result_type(val) == dtypes.float0:
+      return np.zeros((), dtype=bool)
```
It may be that we could put this logic elsewhere, like in device_put handlers or something. Any thoughts?

LenaMartens replied:
put in _ndarray_constant_handler.

mattjj reviewed (approved):
Woohoo! Great work!
